docs/exceptions.md

```diff
@@ -0,0 +1,272 @@
+MUST have a comment that describes its preconditions. When it is called, the
+call MUST have an explanation that assures why these preconditions hold.
+
+For instance, when a programmer calls `Prelude.last xs`, he must add a comment
+at this call site that explains why `xs` is guaranteed to be non-empty, even if
+it seems obvious. A better option is to use functions and types that don't allow
+for errors: instead of `Prelude.last` one can use `NonEmpty.last`.
+
+In case of `zipSameLen` we could use length-indexed vectors:
+
+```
+zipSameLen :: Vec n a -> Vec n b -> Vec n (a, b)
+```
+
+It is always a tradeoff whether to allow programmer mistakes, or do type-level
+trickery to avoid them. The decision process for this is out of scope of this
+document.
+
+### Regular error
+
+Consider a function like `readFile`. It is entirely possible that the path
+passed to it may point to a file that does not exist, and the programmer can't
+do anything to prevent this. Existence of a file is not a property of our code,
+it's the property of the outside world, and we have to deal with all possible
+scenarios.
+
+Another example is parsing user input. We might expect the user to enter a
+number, but we have to consider the case that the user enters something else.
+
+In cases when the erroneous scenarios are out of our control, we consider
+these to be regular errors.
+
+### Identifying error class
+
+Can the execution of the error codepath be excluded by code review and static
+verification (more precise types?)
+
+* Yes: it is a programmer mistake
+* No: it is a regular error
+
+## Is the code 100% pure or potentially impure?
+
+### Impure code
+
+The code is considered impure when it's an `IO` action, a function that returns
+an `IO` action, or similar. The code can be _potentially_ impure when it is
+written in an abstract monad that can be instantiated to `IO`.
+
+Definitely impure:
+
+```
+readFile :: FilePath -> IO String
+```
+
+Potentially impure:
+
+```
+lookupThrow :: MonadThrow m => Map k v -> k -> m v
+```
+
+### Pure code
+
+Pure code is one that does not use `IO` operations. For example:
+
+```
+lookupMaybe :: Map k v -> k -> Maybe v
+```
+
+Notice that while it is conceivable that we have a `Map` of `IO` actions, and
+instantiate `lookupMaybe` to `Map k (IO ()) -> k -> Maybe (IO ())`, the
+instantiation does not inspect/use these `IO` actions, so it's not considered
+potentially impure.
+
+### Identifying code purity
+
+Is there an instantiation of type parameters that would mean that there are
+`IO` actions inspected/used in the code?
+
+* Yes: it is potentially impure code
+* No: it is 100% pure code
+
+## Error handling practices
+
+### Pure code, programmer mistakes
+
+Do *not*:
+
+* do this often, as if it's okay
+* use `undefined`
+* use non-exhaustive pattern matching
+
+Do:
+
+* try to use types to avoid the need in the first place
+* comment extensively (invariants and precondition, reasoning)
```

Comment:
Well, it might be even better to use `LiquidHaskell` for this. I think this is exactly the use case for `LiquidHaskell` — verify some invariants on compilation state.

Reply:
Sure, we could experiment with `LiquidHaskell` in the future. Right now it does not seem mature enough for a project on the scale of Cardano, but I'd like to be proven wrong.